ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionsForSenderInfo.java

```diff
@@ -22,15 +22,15 @@
 import java.util.OptionalLong;
 import java.util.PriorityQueue;
 import java.util.Queue;
-import java.util.TreeMap;
+import java.util.concurrent.ConcurrentSkipListMap;
 import java.util.stream.LongStream;
 
 class TransactionsForSenderInfo {
   private final NavigableMap<Long, PendingTransactions.TransactionInfo> transactionsInfos;
   private final Queue<Long> gaps = new PriorityQueue<>();
 
   TransactionsForSenderInfo() {
-    transactionsInfos = new TreeMap<>();
+    transactionsInfos = new ConcurrentSkipListMap<>();
```

Comment:
If we are replacing with a concurrrrent safe implementation why do we still have sync blocks on the collection?  Was this an accidental include?  Given that we are fixing gaps in a single account set a hard lock seems safe and the overhead on the concurrent skip list may not be worth it especially when using in a non-concurrent fashion.

Reply:
I added this modification because we can have ConcurrentModificationException during this call `createSenderTransactionOrder`

I just reproduced to get the stacktrace :

>java.util.ConcurrentModificationException
	at java.base/java.util.TreeMap$ValueSpliterator.forEachRemaining(TreeMap.java:2901)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.hyperledger.besu.ethereum.core.AccountTransactionOrder.<init>(AccountTransactionOrder.java:32)
	at org.hyperledger.besu.ethereum.eth.transactions.PendingTransactions.createSenderTransactionOrder(PendingTransactions.java:237)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1138)
	at org.hyperledger.besu.ethereum.eth.transactions.PendingTransactions.selectTransactions(PendingTransactions.java:211)
	at org.hyperledger.besu.ethereum.blockcreation.BlockTransactionSelector.buildTransactionListForBlock(BlockTransactionSelector.java:161)
	at org.hyperledger.besu.ethereum.blockcreation.AbstractBlockCreator.selectTransactions(AbstractBlockCreator.java:233)
	at org.hyperledger.besu.ethereum.blockcreation.AbstractBlockCreator.createBlock(AbstractBlockCreator.java:153)
	at org.hyperledger.besu.ethereum.blockcreation.AbstractBlockCreator.createBlock(AbstractBlockCreator.java:126)
	at org.hyperledger.besu.ethereum.blockcreation.BlockMiner.mineBlock(BlockMiner.java:123)
	at org.hyperledger.besu.consensus.clique.blockcreation.CliqueBlockMiner.mineBlock(CliqueBlockMiner.java:49)
	at org.hyperledger.besu.ethereum.blockcreation.BlockMiner.run(BlockMiner.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:830)


But I think keeping hard lock for the `addTransactionToTrack` and `updateGap` it's important so that the info in the queue is in the correct order compared to the reading of the treemap but maybe I missed something

Reply:
Yep.  it needs to stay then.